cmds/webboot/webboot.go

```diff
@@ -10,60 +11,56 @@ import (
 	"path/filepath"
 	"regexp"
 
+	ui "github.com/gizak/termui/v3"
 	"github.com/u-root/webboot/pkg/menu"
 )
 
-// Exec downloads the iso and boot it.
-func (b *BookMarkISO) Exec() error {
-	fPath := filepath.Join("/tmp", b.name)
-	if err := download(b.url, fPath); err != nil {
-		return err
-	}
-	// todo: boot the iso
-	log.Printf("ISO is downloaded at %s", fPath)
-	return nil
-}
-
-// Exec displays a menu of bookmarks
-func (d *DownloadByBookmark) Exec() error {
-	entries := []menu.Entry{}
-	for _, e := range bookmark {
-		entries = append(entries, e)
-	}
-
-	_, err := menu.DisplayMenu("Bookmarks", "Input your choice", entries, d.uiEvents)
-	if err != nil {
-		return err
-	}
+var (
+	v       = flag.Bool("verbose", false, "Verbose output")
+	verbose = func(string, ...interface{}) {}
+)
 
-	return nil
+// exec downloads the iso and boot it.
+func (i *ISO) exec() error {
 	// todo: boot the iso
+	verbose("ISO is at %s\n", i.path)
+	return nil
 }
 
-// Exec asks for link and name, then downloads the iso and boot it.
-func (d *DownloadByLink) Exec() error {
-	link, err := menu.NewInputWindow("Enter URL:", menu.AlwaysValid, d.uiEvents)
-	if err != nil {
-		return err
-	}
+// exec lets user input the name of the iso they want
+// if this iso is existed in the bookmark, use it's url
+// elsewise ask for a download link
+func (d *DownloadOption) exec(uiEvents <-chan ui.Event) (menu.Entry, error) {
 	validIsoName := func(input string) (string, string, bool) {
 		re := regexp.MustCompile(`[\w]+.iso`)
 		if re.Match([]byte(input)) {
 			return input, "", true
 		}
 		return "", "File name should only contain [a-zA-Z0-9_], and should end in .iso", false
 	}
-	filename, err := menu.NewInputWindow("Enter ISO name", validIsoName, d.uiEvents)
+	filename, err := menu.NewInputWindow("Enter ISO name", validIsoName, uiEvents)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	fPath := filepath.Join("/tmp", filename)
-	if err := download(link, fPath); err != nil {
-		return err
+
+	link, ok := bookmarks[filename]
+	if !ok {
+		if link, err = menu.NewInputWindow("Enter URL:", menu.AlwaysValid, uiEvents); err != nil {
+			return nil, err
+		}
 	}
-	// todo: boot the iso
-	log.Printf("ISO is downloaded at %s", fPath)
-	return nil
+
+	fpath := filepath.Join("/tmp", filename)
+	if err = download(link, fpath); err != nil {
```

Comment:
I forgot did we ever end up adding a progress bar for the download?

Reply:
we haven't add this. but I will do some research about how to add it now.